core-plugins/src/main/java/co/cask/hydrator/plugin/common/AvroToStructuredTransformer.java

```diff
@@ -58,42 +43,34 @@ public StructuredRecord transform(GenericRecord genericRecord, Schema structured
       String fieldName = field.getName();
       builder.set(fieldName, convertField(genericRecord.get(fieldName), field.getSchema()));
     }
-
     return builder.build();
   }
 
-  public StructuredRecord transform(StructuredRecord.Builder builder, GenericRecord genericRecord,
-                                    Schema structuredSchema, String skipField)
+  public StructuredRecord.Builder transform(GenericRecord genericRecord, Schema structuredSchema, String skipField)
     throws IOException {
+    StructuredRecord.Builder builder = StructuredRecord.builder(structuredSchema);
     for (Schema.Field field : structuredSchema.getFields()) {
       String fieldName = field.getName();
       if (!fieldName.equals(skipField)) {
         builder.set(fieldName, convertField(genericRecord.get(fieldName), field.getSchema()));
       }
     }
 
-    return builder.build();
+    return builder;
   }
 
-  public StructuredRecord transform(GenericRecord genericRecord, String addFieldKey, String addFieldValue)
-    throws IOException {
-    Schema schema = Schema.parseJson(genericRecord.getSchema().toString());
-    Schema.Field newField = Schema.Field.of(addFieldKey, Schema.of(Schema.Type.STRING));
-    List<Schema.Field> newFields = new ArrayList<>();
-
-    for (Schema.Field field : schema.getFields()) {
-      newFields.add(field);
-    }
-    newFields.add(newField);
-    Schema newSchema = Schema.recordOf(schema.getRecordName(), newFields);
+  public Schema generateStructuredSchema(GenericRecord genericRecord) throws IOException {
```

Comment:
take an avro Schema as the argument instead of a Generic Record and rename it convertSchema